ui/src/components/ContextMenu/index.tsx

```diff
@@ -1,201 +1,90 @@
-  <ContextMenuPrimitive.CheckboxItem
-    ref={ref}
-    className={cn(
-      "relative flex cursor-pointer select-none items-center rounded-sm py-1.5 pl-8 pr-2 text-sm outline-hidden focus:bg-accent focus:text-accent-foreground data-disabled:pointer-events-none data-disabled:opacity-50",
-      className,
-    )}
-    checked={checked}
-    {...props}>
-    <span className="absolute left-2 flex size-3.5 items-center justify-center">
-      <ContextMenuPrimitive.ItemIndicator>
-        <CheckIcon className="size-4" />
-      </ContextMenuPrimitive.ItemIndicator>
-    </span>
-    {children}
-  </ContextMenuPrimitive.CheckboxItem>
-));
-ContextMenuCheckboxItem.displayName =
-  ContextMenuPrimitive.CheckboxItem.displayName;
+import type { Node } from "@flow/types";
+
+export type ContextMenuMeta = {
+  node?: Node;
+  top: number | false;
+  left: number | false;
+  right: number | false;
+  bottom: number | false;
+};
 
-const ContextMenuRadioItem = React.forwardRef<
-  React.ElementRef<typeof ContextMenuPrimitive.RadioItem>,
-  React.ComponentPropsWithoutRef<typeof ContextMenuPrimitive.RadioItem>
->(({ className, children, ...props }, ref) => (
-  <ContextMenuPrimitive.RadioItem
-    ref={ref}
-    className={cn(
-      "relative flex cursor-pointer select-none items-center rounded-sm py-1.5 pl-8 pr-2 text-sm outline-hidden focus:bg-accent focus:text-accent-foreground data-disabled:pointer-events-none data-disabled:opacity-50",
-      className,
-    )}
-    {...props}>
-    <span className="absolute left-2 flex size-3.5 items-center justify-center">
-      <ContextMenuPrimitive.ItemIndicator>
-        <DotFilledIcon className="size-4 fill-current" />
-      </ContextMenuPrimitive.ItemIndicator>
-    </span>
-    {children}
-  </ContextMenuPrimitive.RadioItem>
-));
-ContextMenuRadioItem.displayName = ContextMenuPrimitive.RadioItem.displayName;
+type ContextMenuProps = {
+  items: ContextMenuItemProps[];
+  onClose: () => void;
+  contextMenuMeta: ContextMenuMeta;
+};
 
-const ContextMenuLabel = React.forwardRef<
-  React.ElementRef<typeof ContextMenuPrimitive.Label>,
-  React.ComponentPropsWithoutRef<typeof ContextMenuPrimitive.Label> & {
-    inset?: boolean;
-  }
->(({ className, inset, ...props }, ref) => (
-  <ContextMenuPrimitive.Label
-    ref={ref}
-    className={cn(
-      "px-2 py-1.5 text-sm font-semibold text-foreground",
-      inset && "pl-8",
-      className,
-    )}
-    {...props}
-  />
-));
-ContextMenuLabel.displayName = ContextMenuPrimitive.Label.displayName;
+const ContextMenu: React.FC<ContextMenuProps> = ({
+  items,
+  onClose,
+  contextMenuMeta,
+}) => {
+  const { top, left, right, bottom } = contextMenuMeta;
+  return (
+    <div
+      className="absolute z-50"
+      style={{
+        top: top !== false ? top : undefined,
+        left: left !== false ? left : undefined,
+        right: right !== false ? right : undefined,
+        bottom: bottom !== false ? bottom : undefined,
```

Comment:
If you can update the `handleNodeContextMenu` to return only number values you can avoid this here. 
So doing somehting like 
```
let styles = {};
if(localY < pane.height - 200){
...{
else if(...
...
```
It'll make that code longer, but will only output what is needed